File: cpp/238_product_of_array_except_self.cpp
#include <vector>
using namespace std;

class Solution {
 public:
  vector<int> productExceptSelf(vector<int>& nums) {
    int n = nums.size();
    vector<int> result(n);
    int product = 1;
    for (int i = 0; i < n; ++i) {
      result[i] = product;
      product *= nums[i];
    }

    product = 1;
    for (int i = n - 1; i > -1; --i) {
      result[i] *= product;
      product *= nums[i];
    }
    return result;
  }
};

File: rust/238_product_of_array_except_self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut product = 1;
        let n = nums.len();
        let mut result = vec![0; n];
        for i in 0..n {
            result[i] = product;
            product = product * nums[i];
        }

        product = 1;
        for i in (0..n).rev() {
            result[i] = result[i] * product;
            product = product * nums[i];
        }

        result
    }
}
